### ISSUE TITLE
Questions about scoping rules and PEP 563

### ISSUE THREAD

**Issue by tristanlatr** [microsoft/pyright#4726]
**Describe the current behaviour**
Whether we use `from __future__ import annotations` or not radically changes the scoping rules. I suppose this is designed this way. But `pyright` is not in agreement with `mypy` on this particular topic.

**To Reproduce**

This typechecks correctly with `pyright` (but `mypy` complains about `F` being not a valid type). I'm not saying it shouldn't, just saying it looks like there are some discrepancies in how `mypy` and `pyright` handle this specific (and totally theoretical) case. 

```python
from __future__ import annotations
class C:
  F = b'not a valid type'
  def f(self) -> F:
    return F()

class F:
  ...
```

**Expected behavior**
I'd expect `pyright` to have the same behaviour as `mypy` (or the other way around, I don't care). From the perspective of a  developer of static analysis tools, I don't know which behaviour is the "right way". I've read about all the typing PEPs and couldn't find a clear and definitive answer. 

Thanks

PS, this is not a bug report.

**Comment by erictraut** [comment]
Pyright properly reflects the runtime behavior of annotation evaluation. This behavior changes when `from __future__ import annotations` is enabled. Similarly, the behavior changes when an annotation is enclosed in quotes.

Mypy does not as faithfully reflect the runtime behavior. You can verify this by running the following code:

```python
class C:
    F = b"not a valid type"

    def f(self) -> F:
        return F()

class F:
    ...

print(C.f.__annotations__)
```

The above code outputs:
* "{'return': 'F'}" when PEP 563 is in effect
* {'return': b'not a valid type'} when PEP 563 is not in effect

Pyright's evaluation matches the runtime behavior in both cases. Mypy's evaluation is inconsistent with the runtime behavior when PEP 563 is in effect.

The runtime evaluation order of annotations has changed in the past and is likely to change again (perhaps as soon as Python 3.12), so I strongly recommend that you don't write code that depends on evaluation order. You can avoid this by eliminating symbol name conflicts within your code. If you don't have name conflicts, then this issue doesn't matter.
